typed-racket-lib/typed-racket/typecheck/tc-funapp.rkt

```diff
@@ -55,7 +56,23 @@
                    (list (cons 0 idx) o t))))
 
 (define (tc/funapp f-stx args-stx f-type args-res expected)
-  (match-define (list (tc-result1: argtys (PropSet: argps+ argps-) argobjs) ...) args-res)
+  (define-values (argtys argprops argobjs)
+    (cond
+      [(with-linear-integer-arithmetic?)
+       (for/lists (_1 _2 _3) ([stx (in-syntax args-stx)]
+                              [res (in-list args-res)])
+         (match res
+           [(tc-result1: t (PropSet: p+ p-) (? Empty?))
+            (define o (-id-path (gen-existential-id)))
+            (add-existential-obj! stx o)
```

Comment:
The other option would be to have a function like `single-value` (i.e. that typechecks an expression asserting it returns a single value) that indicates the expression should be assigned a existential object if it does not have a non-empty object, and then that existential object would be what is added to the type table initially (meaning we don't need this function to add it after the fact)